calico/openstack/t_etcd.py

```diff
@@ -34,16 +49,283 @@ def __init__(self, driver, logger):
+                    del ports[endpoint_id]
+
+                elif (endpoint_id not in ports or
+                      hostname != ports[endpoint_id]['binding:host_id']):
+                    LOG.debug("Existing etcd endpoint key is now invalid")
+                    # OpenStack no longer has an endpoint with the ID in the
+                    # etcd key; or it does, but the endpoint has migrated to a
+                    # different host than the one in the etcd key.  In both
+                    # cases the etcd key is no longer valid and should be
+                    # deleted.  In the migration case, data will be written
+                    # below to an etcd key that incorporates the new hostname.
+                    self.client.delete(child.key)
+
+        # Now write etcd data for any endpoints remaining in the ports dict;
+        # these are new endpoints - i.e. never previously represented in etcd
+        # data - or endpoints that have migrated or whose data has changed.
+        for port in ports.values():
+            data = self.port_etcd_data(port)
+            self.client.write(self.port_etcd_key(port), json.dumps(data))
+
+            # Remember the security profile that this port needs.
+            self.needed_profiles.add(data['profile_id'])
+
+    def port_etcd_key(self, port):
+        return "/calico/host/%s/workload/openstack/endpoint/%s" % (
+            port['binding:host_id'],
+            port['id']
+        )
+
+    def port_etcd_data(self, port):
+        # Construct the simpler port data.
+        data = {'state': 'active' if port['admin_state_up'] else 'inactive',
+                'name': port['interface_name'],
+                'mac': port['mac_address'],
+                'profile_id': self.port_profile_id(port)}
+
+        # Collect IPv6 and IPv6 addresses.  On the way, also set the
+        # corresponding gateway fields.  If there is more than one IPv4 or IPv6
+        # gateway, the last one (in port['fixed_ips']) wins.
+        ipv4_nets = []
+        ipv6_nets = []
+        for ip in port['fixed_ips']:
+            if ':' in ip['ip_address']:
+                ipv6_nets.append(ip['ip_address'] + '/128')
+                if ip['gateway'] is not None:
+                    data['ipv6_gateway'] = ip['gateway']
+            else:
+                ipv4_nets.append(ip['ip_address'] + '/32')
+                if ip['gateway'] is not None:
+                    data['ipv4_gateway'] = ip['gateway']
+        data['ipv4_nets'] = ipv4_nets
+        data['ipv6_nets'] = ipv6_nets
+
+        # Return that data.
+        return data
+
+    def port_profile_id(self, port):
+        return '_'.join(port['security_groups'])
+
+    def resync_security_groups(self):
+        # Get all current security groups from the OpenStack database and key
+        # them on security group ID.
+        self.sgs = {}
+        for sg in self.driver.get_security_groups():
+            self.sgs[sg['id']] = sg
```

Comment:
Is this a good idea? Storing the full set of security groups on the box in memory is likely to be a scale bottleneck.